### PR TITLE
Avoid returning clones of pointer types on `JsValue` methods

### PR DESCRIPTION
**Pull request by jedel1043**
This Pull Request creates a new `Ref` struct (design by @HalidOdat) that represents a reference for boxed pointer types (`JsValue`, `JsString`, `JsSymbol`, `JsBigInt`), skipping unnecessary clones.

It changes the following:

- Creates a new `Ref` struct representing a reference to a pointer value.
- Makes methods of `JsValue` return `Ref<{PointerValue}>` when possible. 
- Refactors the codebase to accomodate to the new API.

### CODE REVIEW

File: boa_engine/src/value/mod.rs [hunk 1]
```diff
@@ -681,6 +733,8 @@ mod tests_nan_box {
 
         let bigint = value.as_bigint().unwrap();
 
+        println!("pass!");
```
Comment (Razican) [inline boa_engine/src/value/mod.rs:736]: Is this a leftover comment?

Comment (jedel1043) [reply boa_engine/src/value/mod.rs:736]: Yeah, sorry 😅

File: boa_engine/src/value/mod.rs [hunk 2]
```diff
@@ -575,24 +573,61 @@ pub(crate) unsafe trait PointerType {
...
+    fn deref(&self) -> &Self::Target {
+        &*self.inner
+    }
+}
+
+impl<T: PartialEq> PartialEq<T> for Ref<'_, T> {
+    fn eq(&self, other: &T) -> bool {
+        &*self.inner == other
+    }
+}
```
Comment (HalidOdat) [inline boa_engine/src/value/mod.rs:627]: We should add `#[inline]` so they can be inlined in external crates

**Review by HalidOdat — APPROVED** [review]
This looks good! :)

Great work @jedel1043

**Comment by HalidOdat** [comment]
Lets merge this into `nan-boxing` :)

**Pull request merged**